euphonic/powder.py

```diff
@@ -0,0 +1,279 @@
+    sampling: str = 'golden',
+    npts: int = 1000, jitter: bool = False,
+    energy_bins: np.ndarray = None,
+    scattering_lengths: Union[dict, str] = 'Sears1992'
+) -> Spectrum1D:
+    """Sample the structure factor, averaging over a sphere of constant |q|
+
+    (Specifically, this is the one-phonon inelastic-scattering structure factor
+    as implemented in QpointPhononModes.calculate_structure_factor().)
+
+    Parameters
+    ----------
+
+    fc
+        Force constant data for system
+
+    mod_q
+        scalar radius of sphere from which vector q samples are taken
+
+    dw
+        Debye-Waller exponent used for evaluation of scattering function. If
+        not provided, this is generated automatically over a 20x20x20 q-point
+        mesh.
+
+    temperature
+        Temperature used for Debye-Waller
+
+    sampling
+        Sphere-sampling scheme. (Case-insensitive) options are:
+            - 'golden':
+                Fibonnaci-like sampling that steps regularly along one
+                spherical coordinate while making irrational steps in the
+                other
+
+            - 'sphere-projected-grid':
+                Regular 2-D square mesh projected onto
+                sphere. npts will be distributed as evenly as possible
+                (i.e. using twice as many 'longitude' as 'lattitude' lines),
+                rounding up if necessary.
+
+            - 'spherical-polar-grid':
+                Mesh over regular subdivisions in spherical polar coordinates.
+                npts will be rounded up as necessary in the same scheme as for
+                sphere-projected-grid. 'Latitude' lines are evenly-spaced in z
+
+            - 'spherical-polar-improved':
+                npts distributed as regularly as possible using spherical polar
+                coordinates: 'latitude' lines are evenly-spaced in z and points
+                are distributed among these rings to obtain most even spacing 
+                possible.
+
+            - 'random-sphere':
+                Points are distributed randomly in unit square and projected
+                onto sphere.
+
+    npts
+        Number of samples. Note that some sampling methods have constraints on
+            valid values and will round up as appropriate.
+
+    jitter
+        For non-random sampling schemes, apply an additional random
+            displacement to each point.
+
+    energy_bins
+        Preferred energy bin edges. If not provided, will setup 1000 bins (1001
+        bin edges) from 0 to 1.05 * [max energy]
+
+    scattering_lengths
+        Dict of neutron scattering lengths labelled by
+        element. If a string is provided, this selects coherent scattering
+        lengths from reference data by setting the 'label' argument of the
+        euphonic.util.get_reference_data() function.
+
+    Returns
+    -------
+    Spectrum1D
+
+    """
+
+    if isinstance(scattering_lengths, str):
+        import euphonic.util
```

Comment:
Is there a reason behind importing here instead of at the top of the file?

Reply:
There's a good story here: I've been running into import loops as I've added more type hints. Mostly this has hit me on the spectrum plot branch, but clearly that developed a habit that sneaked in here. Basically the problem is

- top-level `euphonic.__init__` imports classes from nearly every module
- a lot of these classes import functionality from `euphonic.util`
- `euphonic.util` tries to import classes for its type hints
- but we're already trying to import that class...
- BOOM

Deferring imports from `euphonic.util` breaks this loop. It's pretty ugly though, I think a better approach will be to wrap class names in strings in `euphonic.util` to defer/prevent their import there.
In hindsight this is really a consequence of keeping a miscellaneous `util` module rather than establishing those features within a clearer heirarchy!

In any case I don't think it's still a problem on _this_ branch, so will see about taking it out.

Reply:
Yeah I've had trouble in the past with this... I was also thinking about separating out util at some point because its just starting to become a very long list of functions. Maybe `__init__` could do with a rethink too.

Reply:
Issue opened #90